File: python/play.py
#!/usr/bin/env python3

"""Play a file to an E1/T1 timeslot using a Corelatus GTH
"""

# Copyright (c) 2020–2025, Corelatus AB
# All rights reserved.
#
# Licensed under the BSD 3-Clause License. See the LICENSE file
# in the project root for full license information.

import sys
from sys import argv, stderr
import gth.apilib

def _usage():
    stderr.write("""
play <hostname> <span> <timeslot> <filename>

  <hostname>: the hostname or IP address of a GTH
      <span>: the name of an E1/T1, e.g. 1A or 4D on a GTH 2.x
  <timeslot>: 1--31 on an E1 or 1--24 on a T1
  <filename>: which file to write the data to. - means stdin

Typical invocation: ./play 172.16.1.10 1A 16 signalling.raw
""")

def _play(host, span, timeslot, file):
    api = gth.apilib.API(host)
    api.warn_if_l1_dead(span)

    _player_id, data = api.new_player(span, timeslot)
    octets_sent = 0

    # Stream out the message
    while True:
        buffer = file.read(1000)
        octets_sent += len(buffer)
        if buffer == b"":
            break
        data.sendall(buffer)

    data.close()
    stderr.write(f"All done, sent {octets_sent} octets\n")

    # Use 13 second timeout because file is about 12 seconds long
    while api.next_event(13.0)[1] != 'message_ended':
        pass

    api.bye()

def main():
    """entry point"""
    if len(sys.argv) != 5:
        _usage()
        sys.exit(-1)

    # We only check the number of arguments. If the GTH doesn't like
    # the contents, it'll say so.

    if argv[4] == "-":
        _play(argv[1], argv[2], int(argv[3]), sys.stdin.buffer)
    else:
        with open(argv[4], "rb") as f:
            _play(argv[1], argv[2], int(argv[3]), f)

main()

File: perl/play.pl
#!/usr/bin/perl -Igth_control/lib/
#
# Title: Play a file to an E1/T1 timeslot using a Corelatus GTH
# Author: Matthias Lang (matthias@corelatus.se)
#

use 5.008;         # Probably works with earlier versions too.
use strict;
use warnings;
use gth_control;
use Data::Dumper;

sub usage() {
    print("
play [-v] <hostname> <span> <timeslot> <filename>

  <hostname>: the hostname or IP address of a GTH
      <span>: the name of an E1/T1, e.g. 1A or 4D on a GTH 2.x
  <timeslot>: 1--31 on an E1 or 1--24 on a T1
  <filename>: which file to write the data to. - means stdin
          -v: verbose, for debugging

Typical invocation: ./play 172.16.1.10 1A 16 signalling.raw
");
}

# Check that a given PCM is in a state where it could give useful data.
# That means in 'OK' or 'RAI' status.
sub warn_if_l1_dead {
    my ($api, $span) = @_;

    my $status = $api->query_resource("pcm$span", "status");

    if ($status eq "OK") {
    } elsif ($status eq "RAI") {
    } elsif ($status eq "disabled") {
	print STDERR "Warning: pcm$span is disabled. The GTH won't actually emit any data.\n".
	    "Hint: enable L1 with enable_l1.pl\n";
    } else {
	print STDERR "Warning: pcm$span status is $status\n" .
	    "Chances are the other end of your E1 isn't plugged in (or enabled)\n";
    }
}

sub play {
    my ($host, $span, $timeslot, $file, $verbose) = @_;
    my $api = new gth_control($host, $verbose);
    warn_if_l1_dead($api, $span);

    my ($player_id, $data) = $api->new_player($span, $timeslot);
    my $octets_sent = 0;
    my $buffer = "";

    # Stream out the message
    while (read($file, $buffer, 1000)) {
	$octets_sent += length($buffer);
	print $data $buffer;
    }

    close($file);
    close($data);

    # Check for a message-ended event. Since we only start one player
    # we don't need to check the ID on the event.
    while (1) {
	my $event = $api->next_event();
	my $message_ended = $event->{"event"}[0]->{"message_ended"};
	!defined $message_ended || last;
    }

    print STDERR "All done, sent $octets_sent octets\n";

    $api->bye();
}

# Entry point
my $verbose = 0;
if ($ARGV[0] eq "-v") {
    $verbose = 1;
    shift @ARGV;
}

($#ARGV == 3) || usage() && die();

my $host = $ARGV[0];
my $span = $ARGV[1];
my $file;

if ($ARGV[3] eq "-") {
    $file = *STDIN;
} else {
    open($file, $ARGV[3]) || die("can't open $ARGV[3]");
}

play($ARGV[0], $ARGV[1], $ARGV[2], $file, $verbose);

